Bugzilla/API/V1/Github.pm

```diff
@@ -164,7 +171,167 @@ sub pull_request {
   return $self->render(json => {error => 0, id => $attachment->id});
 }
 
-sub verify_signature {
+sub push_comment {
+  my ($self) = @_;
+  my $template = Bugzilla->template;
+  Bugzilla->usage_mode(USAGE_MODE_MOJO_REST);
+
+  # Return early if push commenting is not allowed
+  return $self->code_error('github_push_comment_disabled')
+    if !Bugzilla->params->{github_push_comment_enabled};
+
+  # Return early if not a push or ping event
+  my $event = $self->req->headers->header('X-GitHub-Event');
+  if (!$event || ($event ne 'push' && $event ne 'ping')) {
+    return $self->code_error('github_push_comment_not_push');
+  }
+
+  # Verify that signature is correct based on shared secret
+  if (!$self->_verify_signature) {
+    return $self->code_error('github_push_comment_mismatch_signatures');
+  }
+
+  # If event is a ping and we passed the signature check
+  # then return success
+  if ($event eq 'ping') {
+    return $self->render(json => {error => 0});
+  }
+
+  # Parse push commit title for bug ID
+  my $payload = $self->req->json;
+  if ( !$payload
+    || !$payload->{ref}
+    || !$payload->{repository}
+    || !$payload->{repository}->{full_name}
+    || !$payload->{pusher}
+    || !$payload->{pusher}->{name}
+    || !$payload->{commits})
+  {
+    return $self->code_error('github_push_comment_invalid_json');
+  }
+
+  my $ref        = $payload->{ref};
+  my $repository = $payload->{repository}->{full_name};
+  my $pusher     = $payload->{pusher}->{name};
+  my $commits    = $payload->{commits};
+
+  # Return success early if there are no commits or the ref is not a branch
+  if (!@{$commits} || $ref !~ /refs\/heads\//) {
+    return $self->render(json => {error => 0});
+  }
+
+  # Keep a list of bug ids that need to have comments added.
+  # We also use this for sending email later.
+  # Use a hash so we don't have duplicates. If multiple commits
+  # reference the same bug ID, then only one comment will be added
+  # with the text combined.
+  # When the comment is created, we will store the comment id to
+  # return to the caller.
+  my %update_bugs;
+
+  # Create a separate comment for each commit
+  foreach my $commit (@{$commits}) {
+    my $message = $commit->{message};
+    my $url     = $commit->{url};
+
+    if (!$url || !$message) {
+      return $self->code_error('github_pr_invalid_json');
+    }
```

Comment:
Can we log and/or return information about what is wrong with the JSON here?

Reply:
I will add a logging call that we can use for debugging if needed but I am reluctant to return the Perl error message verbatim so as to not open a potential security issue in the future. They can contact a BMO admin to help with debugging if needed.

Reply:
> I will add a logging call that we can use for debugging if needed but I am reluctant to return the Perl error message verbatim so as to not open a potential security issue in the future. They can contact a BMO admin to help with debugging if needed.

Scratch that. I am not failing if the JSON parser fails, only if a required key(s) is missing. I will update the error to be more descriptive if possible.